// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

syntax="proto2";

package brpc.policy;

option cc_generic_services = true;
option java_generic_services = true;
option java_package="com.brpc.policy";
option java_outer_classname="WebSocketProto";

enum WebSocketOpcode {
    CONTINUE_FRAME = 0;
    TEXT_FRAME = 1;
    BINARY_FRAME = 2;
    CONNECTION_CLOSE_FRAME = 8;
    PING_FRAME = 9;
    PONG_FRAME = 10;
}

message WebSocketHeader {
    required int64 message_length = 1;
    optional int32 request_id = 2;
    required int32 fin = 3;
    required WebSocketOpcode opcode = 4;
    optional bytes mask_key = 5;
}

message WebSocketRequest {
    required WebSocketHeader header = 1;
    optional bytes message = 2;
}

message WebSocketResponse {
    required WebSocketHeader header = 1;
    optional bytes message = 2;
}

service WebSocketService {
    rpc default_method(WebSocketRequest) returns (WebSocketResponse);
}
